### ISSUE TITLE
Allow a plugin to override the Request used as cache key

### ISSUE THREAD

**Issue by jeffposnick** [GoogleChrome/workbox#1364]
**Library Affected**:
*workbox-core*

Extracting a feature request from https://github.com/GoogleChrome/workbox/issues/1354#issuecomment-373188977, @josephliccini writes:

> Just brainstorming here.. It might make sense to have a plugin / callback that enables the developer to set their custom Request key logic.

The use case, as explained in the bug, is scenarios in which you don't want the full `Request`, including headers, saved to disk via the Cache Storage API. (It's worth following-up out of band as to whether that's actually a concern before we potentially move ahead with this.)

**Comment by gauntface** [comment]
Sorry, but I'm not fully understanding the issue, use case and desires of this issue, specifically:

1. How is workbox preventing custom logic to do this?
1. Is this really a common enough use case to support in Workbox directly?

**Comment by josephliccini** [comment]
I was able to achieve the functionality desired like this: 
```js
const removeRequestInformationPlugin = {
    cacheDidUpdate: async ({ url, cacheName }) => {
        // Get the result that was just placed into the cache
        const cache = await caches.open(cacheName);
        const response = await cache.match(url);

        // Re-set the cache without any headers or metadata
        await cache.put(url, response);
    }
};

workboxSW.router.registerRoute(/\/mock\//, workboxSW.strategies.staleWhileRevalidate({
    cacheName: 'mock-api',
    cacheExpiration: {
        maxAgeSeconds: 3600
    },
    cacheableResponse: {
        statuses: [
            200
        ]
    },
    plugins: [
        removeRequestInformationPlugin
    ]
}), 'GET')
```

I think that this might be OK for now, but I think a better implementation would be something along the lines of a lifecycle hook / plugin callback that would be `requestWillBeUsedAsKey`, that would allow for custom key modification.  

An option would be if we could use the `cacheWillUpdate` and modify the request there, and then that request would be used as the key.  But I did get an exception that disallowed multiple `cacheWillUpdate` plugins..

**Comment by jeffposnick** [comment]
This was fixed back in #1990.
